Theme fixes

gzdunek opened the pull request:
This PR mainly adjusts Connect to the updated dark theme. Some of the changes affect also the WebUI, so @rudream I'd be grateful if you could take a look at the non-Connect commits (up to this one https://github.com/gravitational/teleport/commit/80f2124012e93233e240d35699ce73f9cf327305).

Justification for the changes that affect WebUI:
- ~https://github.com/gravitational/teleport/commit/1fe1d7baecba3991d58d6f822cf72485d613cb3a - the pagination icon is barely visible when disabled~ 
- https://github.com/gravitational/teleport/commit/f16209d5985ff7f0063914c7ee298abed76e1e60 - I don't know if there was any design for that, but here https://www.figma.com/file/Gpjs9vjhzUKF1GDbeG9JGE/Application-Design-System?node-id=7102%3A34980&t=r6OqTad9lQjTeVqw-1 Kenny used a different color. IMO looks better.
- https://github.com/gravitational/teleport/pull/24883/commits/a27ec5ef02878beceb2ca49e5296fde9818a9e33 - on Firefox placeholders  are a little more transparent https://developer.mozilla.org/en-US/docs/Web/CSS/::placeholder#opaque_text
- https://github.com/gravitational/teleport/pull/24883/commits/f3db99dbb93a6e88f859a7ca32a1d81d4696cd96 - I don't know if there was a design for it too, but `secondary` text seems to dark for list items (and doesn't meet accessibility requirements in Chrome dev tools). 

@ravicious if something in Connect doesn't look right, feel free to commit fixes.

<img width="1264" alt="image" src="https://user-images.githubusercontent.com/20583051/233370637-625e11dc-407d-4b64-bb5f-d1f915fc4144.png">

gzdunek committed:
Make pagination icons a little more visible

gzdunek committed:
Use different color for `tr` border

gzdunek committed:
Use `opacity: 1` for placeholders

gzdunek committed:
Include `borderRadius` in `StyledPanel` to fix rounded corners in Connect custom tables

gzdunek committed:
Use `colors.text.primary` for items in `MenuLogin`

gzdunek committed:
Adjust Connect theme to the updated dark theme

gzdunek committed:
Remove unused component

gzdunek committed:
Update snapshots

ravicious commented on web/packages/design/src/TextArea/TextArea.tsx:
```diff
@@ -28,7 +28,7 @@ export interface TextAreaProps extends React.ComponentPropsWithRef<'textarea'> {
 
 export const TextArea: React.FC<TextAreaProps> = styled.textarea`
```
What's the reason behind changes to TextArea? These strike me as substantial and they're not included in the PR description.

gzdunek replied:
For some reason, when I originally added `TextArea` I didn't include all these styles in that component. It feels weird to add them in every place where `TextArea` is used. 
Now `TextArea` is more similar to `Input` which includes all styles.
Also, this component is used only in our feedback form.

ravicious commented on web/packages/teleterm/src/ui/Search/SearchBar.tsx:
```diff
@@ -103,14 +103,12 @@ function SearchBar() {
-        background: ${props => props.theme.colors.levels.sunkenSecondary};
-        border: 1px ${props => props.theme.colors.action.disabledBackground}
-          solid;
+        background: inherit;
```
This line can be removed as it doesn't do anything AFAIK.

ravicious commented on web/packages/teleterm/src/ui/Search/SearchBar.tsx:
```diff
@@ -103,14 +103,12 @@ function SearchBar() {
-        border: 1px ${props => props.theme.colors.action.disabledBackground}
-          solid;
+        background: inherit;
+        border: 1px ${props => props.theme.colors.buttons.border.border} solid;
```
I don't think this was introduced by this PR specifically and I don't think we necessarily need to fix this right now. I just want to point out something we should be aware of in the future.

We have five different elements in the top bar and almost every one of them behaves in a different way when it comes to how the border is handled when the element is active.


https://user-images.githubusercontent.com/27113/233377044-ffaf1cab-d325-4b9e-92e3-38698d7102d0.mov

gzdunek replied:
Definitely we should change the search input border in the cluster selector.
It would be good to get rid of the connections icon background, but the activity indicator looks weird then.

ravicious replied:
> It would be good to get rid of the connections icon background, but the activity indicator looks weird then.

Yeah, and we're going to remove this whole UI element soon anyway.

ravicious commented on web/packages/teleterm/src/ui/Search/pickers/PickerContainer.ts:
```diff
@@ -24,10 +24,11 @@ export const PickerContainer = styled.div`
+  box-shadow: ${props => props.theme.boxShadow[1]};
   border-radius: ${props => props.theme.radii[2]}px;
-  border: 1px solid ${props => props.theme.colors.action.hover};
+  // border in the same color as the background to prevent layout shifting
```
Could you go into more detail here? I was able to understand what this does by removing the style but I feel like the comment alone doesn't tell the whole story. As far as I understand, this has to do with PickerContainer wrapping the input when the search bar is open, so we have to use a border of the same width to keep the search bar layout from shifting when switching between open and closed state, right?

gzdunek replied:
That's the exact reason, I will include it in the comment.

ravicious commented on web/packages/teleterm/src/ui/Search/pickers/ResultList.tsx:
```diff
@@ -139,27 +139,26 @@ export const NonInteractiveItem = styled.div`
   padding: ${props => props.theme.space[2]}px;
   color: ${props => props.theme.colors.text.contrast};
-  background: ${props => props.theme.colors.levels.surface};
+  background: inherit;
```
This can be removed.

ravicious commented on web/packages/teleterm/src/ui/Search/pickers/ResultList.tsx:
```diff
@@ -139,27 +139,26 @@ export const NonInteractiveItem = styled.div`
   &:hover,
   &:focus {
-    background: ${props => props.theme.colors.levels.elevated};
+    background: ${props => props.theme.colors.spotBackground[0]};
```
@rudream What's the "semantic" difference between `levels.popout` and `spotBackground`, between `spotBackground[0]` and `spotBackground[2]`?

I feel like we should use the elevation scheme all the way through, `spotBackground` feels like an escape hatch for when it was inconvenient to use `levels`. 🤔

@gzdunek But I think in this case we could just use `levels.popout` instead of `spotBackground[0]` here, couldn't we?

roraback replied:
Levels are for large areas and are solid colors. 

Spot backgrounds are for smaller elements, are mostly transparent so they can add show a change within elements that sit atop the current level. As such, they are most frequently used for interactivity, because their transparency allows them to work across every single level. 

Levels should not be used to designate interactivity—they're the broad backgrounds used to hold other elements. If we used levels for interactivity, we would need to have unnecessarily complex logic for calibrating interactive effects for things like minimal buttons depending on what level we're placing them in. With spot backgrounds, we can say (for example) that a minimal button always gets spot[0] for a hover state and spot[1] for an active state, regardless of the level.

See [the Figma documentation for Bloomberg](https://www.figma.com/file/Gpjs9vjhzUKF1GDbeG9JGE/Application-Design-System?node-id=7020%3A33956&t=bHgvlFIZDHEtaZHl-1) for a glimpse how this works in practice.

gzdunek replied:
Thanks for the detailed explanation Kenny!

It looks like the included link is incorrect :)

ravicious replied:
> It looks like the included link is incorrect :)

It's correct, I think Kenny meant "look at _our_ Figma documentation we made for a customer". ;)

gzdunek replied:
But by doing this we will apply hover also to the input from the "open state", I've just tried that and it doesn't look good :/

ravicious commented on web/packages/teleterm/src/ui/StatusBar/ShareFeedback/ShareFeedbackFormFields.tsx:
```diff
@@ -76,19 +75,8 @@ export function ShareFeedbackFormFields({
-        &:active {
-          border: 1px solid ${theme.colors.text.secondary};
-        }
+            background: inherit;
```
This is to work around the fact that form elements are not aware of the surface level they're used at, right?

@rudream did you run into similar problems in the Web UI? This feels like an interesting problem to take care of in the future.

This is the background of the textarea that is shown when you click on the little chat icon in the bottom right corner.

ravicious replied:
Well, maybe it's not even that interesting. If we want those field forms to be flat, they could just not set the background color by themselves and instead inherit it, as shown here.

gzdunek replied:
I think it makes sense to inherit parent color, but I was afraid to change it for `Input` which is used in a lot of places :/

ravicious commented on web/packages/teleterm/src/ui/ThemeProvider/theme.ts:
```diff
@@ -31,12 +31,12 @@ const colors = {
   },
 
-  brand: '#512FC9',
+  brand: '#9F85FF',
```
Shouldn't we import those directly from the dark theme instead of copying them here? In the end, shouldn't the Web UI and Connect use the same themes? cc @rudream

ravicious commented on web/packages/teleterm/src/ui/ThemeProvider/theme.ts:
```diff
@@ -31,12 +31,12 @@ const colors = {
     surface: '#222C59',
     surfaceSecondary: '#182047',
```
@rudream What's the story with those `levels.*Secondary` colors, did they get removed from the Web UI themes at some point? @gzdunek We should probably do the same thing that the Web UI did with them.

rudream replied:
Yes, I removed them since they were not part of the figma design system.

ravicious commented on web/packages/teleterm/src/ui/TopBar/Clusters/ClusterSelector/ClusterSelector.tsx:
```diff
@@ -78,15 +78,6 @@ const Container = styled.button`
   &:focus {
     opacity: 1;
-    border-color: ${props => props.theme.colors.light};
+    background: ${props => props.theme.colors.spotBackground[0]};
```
Another candidate to get replaced with `levels.popout`.

rudream replied:
Since this is to highlight a focused state, we should use a `spotBackground`

ravicious commented:
I'll try out those changes tomorrow, so far I've been just going through the code and used the search bar.

roraback commented on web/packages/design/src/DataTable/Pager/StyledPager.tsx:
```diff
@@ -39,7 +39,7 @@ export const StyledArrowBtn = styled.button`
     cursor: default;
     ${Icon} {
-      opacity: 0.1;
+      opacity: 0.2;
```
Oh, this must have been missed. I'm inspecting the active arrow, and it is only .5 opacity; we should have a minimum "muted" white at .54 opacity, and disabled at .36 opacity. I would go with the "slightly muted" setting of .72 for the active icon and .36 for the disabled one...

But now that I'm thinking about it, are these custom-implemented icon buttons? Is there a reason these aren't leveraging our icon button component?

gzdunek replied:
This is definitely a good place, it looks like we are missing those colors. I can add them.

rudream replied:
Ah yes I had missed the new text opacities because I hadn't noticed that they had changed, sorry about that. We should update them. I notice though that in the theme design guidelines we don't have a `placeholder` text opacity, which we use in all of our inputs, could you add one please @roraback?

gzdunek replied:
I reverted this change in the PR, I'll add it in another PR as I found one other thing that we can improve in the table.

gzdunek committed:
Remove `surfaceSecondary` and `sunkenSecondary` colors

gzdunek committed:
Remove unneeded `inherit`

gzdunek committed:
Do not hardcode bg color in `TextArea`

gzdunek committed:
Expand comment

gzdunek committed:
Simplify the look of top bar elements

gzdunek committed:
Remove unused component

gzdunek committed:
Remove `text.contrast`

gzdunek committed:
Use the same dark theme for WebUI and Connect

gzdunek committed:
Revert "Make pagination icons a little more visible"

This reverts commit 1fe1d7baecba3991d58d6f822cf72485d613cb3a.

gzdunek committed:
Add shadow for the tabs

gzdunek commented:
I applied all your comments, so now:
- we don't have `*Secondary` colors anymore
- we don't have `text.contrast` color anymore
- Connect doesn't define its own colors (which were nearly 1:1 copied from the WebUI theme), but instead imports `darkTheme` 
- I adjusted the assumed roles bar to look more like in WebUI 
- I cleaned up the look of the top bar menus (connections and leaf clusters) by limiting the number of used colors 

I feel that the application has gained much more consistency!

-------------
I noticed that there is a problem with lack of separation between an inactive tab and a table when the view is scrolled down (because they use the same color now): 
![image](https://user-images.githubusercontent.com/20583051/234527883-6a00455f-3b02-4a60-93c6-a307f2857348.png)

Kenny's suggestion was to add a drop shadow:
>We may be able to come up with a more elegant solution later, but since I’m currently working on design changes to the resources view, let’s go with the simple solution of adding some shadow:
<img width="1259" alt="image" src="https://user-images.githubusercontent.com/20583051/234528756-d8173935-75ed-4a2a-946e-ade6caf18f00.png">

I like the flat design more, but I don't have any better idea how to fix this. However, if it's a temporary solution then maybe it's ok?

ravicious commented on web/packages/teleterm/src/ui/Tabs/Tabs.tsx:
```diff
@@ -69,7 +69,19 @@ export function Tabs(props: Props) {
+      typography="h5"
+      bold
+      {...styledProps}
+      css={`
```
Any reason these new properties are defined within a `css` prop instead of being defined on `StyledTabs` directly?

gzdunek replied:
Ah no reason, fixed.

ravicious reviewed (commented):
I'm going to look through the app in a sec, just wanted to leave a comment about the shadows.

ravicious commented on web/packages/teleterm/src/ui/Tabs/Tabs.tsx:
```diff
@@ -69,7 +69,19 @@ export function Tabs(props: Props) {
+      css={`
+        position: relative;
+        z-index: 1;
+        box-shadow: 0px 1px 10px 0px rgba(0, 0, 0, 0.12),
```
The shadow covers a small portion of the shell content. Perhaps changing the shell component to add just a little bit of space between the tab and the shell would make the shadow not cover the shell?
 
![shell tab](https://user-images.githubusercontent.com/27113/234600133-f2241639-df35-4574-bf0f-adf66267f222.png)

gzdunek replied:
Done

ravicious commented on web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx:
The main header in `DocumentGateway`, "Database Connection", is less visible than the smaller headers. They should be the same color, I think.

![gateway tab](https://user-images.githubusercontent.com/27113/234608093-8c89cd11-420c-4db6-bd2e-f00bdde90854.png)

gzdunek replied:
Done

ravicious commented on web/packages/teleterm/src/ui/TopBar/TopBar.tsx:
```diff
@@ -49,11 +49,12 @@ export function TopBar() {
   height: 56px;
   align-items: center;
   justify-content: space-between;
+  z-index: 2; // minimally higher z-index than the one defined in StyledTabs, so that its drop-shadow doesn't cover the TopBar 
```
Is it possible to manipulate the box-shadow in a way where it doesn't appear at the top of the div so that we don't have to manipulate z-index to fix this?

gzdunek replied:
I've tried that, but without much success (and our shadow contains 3 sub-shadows which makes it harder).

ravicious commented on web/packages/teleterm/src/ui/ThemeProvider/theme.ts:
```diff
@@ -15,155 +15,15 @@ limitations under the License.
 const theme = {
   name: 'dark',
-  colors,
+  colors: darkTheme.colors,
```
This is super nice. I did try replacing it with the light theme and it doesn't look 100% right, which suggests that some of our components make the wrong assumptions about the theme. But we can take care of these problems when the time comes to add the light theme to Connect.

ravicious reviewed (commented):
Love to see Connect just directly importing the colors from the dark theme. Great job @gzdunek and @rudream!

gzdunek committed:
Merge branch 'master' into gzdunek/theme-updates

# Conflicts:
#	e
#	web/packages/design/src/Input/Input.jsx
#	web/packages/design/src/Menu/MenuItem.jsx
#	web/packages/shared/components/FieldInput/__snapshots__/FieldInput.test.tsx.snap
#	web/packages/shared/components/FormPassword/__snapshots__/FormPassword.test.tsx.snap
#	web/packages/shared/components/MenuLogin/MenuLogin.tsx
#	web/packages/teleport/src/Account/ManageDevices/AddDevice/__snapshots__/AddDevice.story.test.tsx.snap
#	web/packages/teleport/src/Welcome/NewCredentials/__snapshots__/NewCredentials.story.test.tsx.snap
#	web/packages/teleport/src/components/FormLogin/__snapshots__/FormLogin.story.test.tsx.snap
#	web/packages/teleterm/src/ui/DocumentCluster/ClusterResources/ClusterSearch.tsx
#	web/packages/teleterm/src/ui/DocumentCluster/ClusterResources/MenuLoginTheme.tsx
#	web/packages/teleterm/src/ui/QuickInput/QuickInput.tsx
#	web/packages/teleterm/src/ui/QuickInput/QuickInputList/QuickInputList.tsx
#	web/packages/teleterm/src/ui/Search/SearchBar.tsx
#	web/packages/teleterm/src/ui/Search/pickers/PickerContainer.ts
#	web/packages/teleterm/src/ui/Search/pickers/ResultList.tsx
#	web/packages/teleterm/src/ui/StatusBar/ShareFeedback/ShareFeedbackFormFields.tsx
#	web/packages/teleterm/src/ui/Tabs/TabItem.tsx
#	web/packages/teleterm/src/ui/ThemeProvider/theme.ts
#	web/packages/teleterm/src/ui/TopBar/Clusters/ClustersFilterableList/ClusterItem.tsx
#	web/packages/teleterm/src/ui/components/FilterableList/FilterableList.tsx
#	web/packages/teleterm/src/ui/components/ListItem.tsx
#	web/packages/teleterm/src/ui/components/Table.tsx

gzdunek committed:
Post-merge fixes

gzdunek committed:
Do not use Arial on custom buttons

gzdunek committed:
Revert snapshot changes

gzdunek committed:
Fix colors in `ActionPicker`

gzdunek committed:
Apply hover styles directly on `SearchBar` input

gzdunek committed:
Use white color for "Database Connection" header

gzdunek committed:
Add shadow directly to `StyledTabs`

gzdunek committed:
Run prettier

gzdunek committed:
Update e

gzdunek committed:
Add some top padding to the terminal

gzdunek commented on web/packages/teleterm/src/ui/Search/SearchBar.tsx:
```diff
@@ -103,14 +103,11 @@ function SearchBar() {
-          color: ${props => props.theme.colors.levels.contrast};
-          background: ${props => props.theme.colors.levels.surface};
+        &:hover > input {
+          background: ${props => props.theme.colors.spotBackground[0]};
```
Done

ravicious commented on web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx:
```diff
@@ -111,6 +111,7 @@ export function DocumentTerminal(props: Props & { visible: boolean }) {
       visible={visible}
       flexDirection="column"
       pl={2}
+      pt={1}
```
Looks good but let's add a comment explaining what the top padding is for.

ravicious commented on web/packages/teleterm/src/ui/Search/pickers/ActionPicker.story.tsx:
```diff
@@ -59,6 +59,7 @@ export const Items = (props: { maxWidth: string }) => {
           max-width: ${maxWidth};
           min-width: 0;
           flex: 1;
+          background-color: ${props => props.theme.colors.levels.elevated};
```
Let's replace `<Flex gap={4}>` above with `<Flex gap={4} alignItems="flex-start">` to get rid of the second column stretching all the way down.

ravicious reviewed (approved):
Left some comments about some minor issues but the rest looks good!

gzdunek committed:
Review fixes

ravicious commented on web/packages/teleterm/src/ui/Search/SearchBar.tsx:
```diff
@@ -108,18 +108,18 @@ function SearchBar() {
       {!isOpen && (
         <>
-          <Input {...defaultInputProps} />
+          <Input {...defaultInputProps} css={`
```
Though what's wrong with defining the background color the way it was, on the flex itself?

https://github.com/gravitational/teleport/blob/abcd52764180199d9f838cc130c19b9157bca8b8/web/packages/teleterm/src/ui/Search/SearchBar.tsx#L114-L117

gzdunek replied:
When I added the semi-transparent hover on the div level, it was brighter than it should (like alpha was 0.14 not 0.07). I don't know why this happened, the input inherited background from the parent. For some reason hover on the div and the input looked a bit different.
But now I found a way to fix it, I only need to set transparent background for the input.

gzdunek committed:
Move SearchBar hover to the Flex element

rudream reviewed (approved):
LGTM

gzdunek committed:
Table-related improvements  (#25333)

* Use a better icon for the reverse tunnel cell

* Use `ButtonIcon` for pagination icons

* Fix rendering table border on Safari

* Manually add a visual separator between the table and the element below it

* Add bottom padding to resource tables in Connect

* Update snapshots

gzdunek committed:
Merge branch 'master' into gzdunek/theme-updates

# Conflicts:
#	e
#	web/packages/teleport/src/Audit/__snapshots__/Audit.story.test.tsx.snap

gzdunek committed:
Update snapshots

gzdunek committed:
Revert e
